node/silkworm/db/mdbx.cpp

```diff
@@ -162,36 +162,43 @@ ::mdbx::cursor_managed open_cursor(::mdbx::txn& tx, const MapConfig& config) {
     return tx.open_cursor(open_map(tx, config));
 }
 
+
 PooledCursor::PooledCursor(::mdbx::txn& tx, const MapConfig& config) {
-    cursor_ = cursors_pool_.acquire();
-    if (!cursor_) {
-        cursor_ = std::make_unique<::mdbx::cursor_managed>();
-    }
+    if (cursors_.empty()) {
+        handle_ = ::mdbx_cursor_create(nullptr);  // create new cursor
+    } else {
+        handle_ = cursors_.back().handle_;  // steal the handle of previously used cursor
+        cursors_.pop_back();
+    } 
```

Comment:
This has a problem:
1) On very first creation of a PooledCursor we create a new one.
2) When that pooled cursor gets destructed its handle is returned to the cache

Now picture this: a new function requires two PooledCursors

1) First one (A) is created by copying handle from the cache
2) pop_back() invokes the dtor of the cached object **but** the object dtor repushes itself in the cache
3) At this stage we have one PooledCursor plus itself in the cache
4) A second PooledCursor (B) is created ... this gets also copied from the cache and eventually bound to a new map
5) Now PooledCursor A is also bound to the same map of PooledCursor B (the share the same ptr to handle) and its position invalidated but in higher code we don't know it.

Reply:
Thank you thank you! I was having a hard time finding the issue. This should be easily fixed.

Reply:
Thanks again @AndreaLanfranchi , I just checked in the fix and I believe the code is safe now. The node_test pass. Basically we want to make sure we always move (not copy) the PooledCursor, so the ownership of the internal pointer remains unique.

Reply:
At family dinner now ... mind if I review tomorrow ?

Reply:
No problem of course, enjoy your family dinner. Italian food is so delicious!